plugin/plugin.go

```diff
@@ -57,72 +56,73 @@ type CLIPlugin struct {
 	var resp proto.GetMetadataResponse
-	err := run(ctx, p.name, p.path, executor, req, &resp)
+	err := run(ctx, p.name, p.path, req, &resp)
 	return &resp, err
 }
 
 func (p *CLIPlugin) DescribeKey(ctx context.Context, req *proto.DescribeKeyRequest) (*proto.DescribeKeyResponse, error) {
 	var resp proto.DescribeKeyResponse
-	err := run(ctx, p.name, p.path, executor, req, &resp)
+	err := run(ctx, p.name, p.path, req, &resp)
 	return &resp, err
 }
 
 func (p *CLIPlugin) GenerateSignature(ctx context.Context, req *proto.GenerateSignatureRequest) (*proto.GenerateSignatureResponse, error) {
 	var resp proto.GenerateSignatureResponse
-	err := run(ctx, p.name, p.path, executor, req, &resp)
+	err := run(ctx, p.name, p.path, req, &resp)
 	return &resp, err
 }
 
 func (p *CLIPlugin) GenerateEnvelope(ctx context.Context, req *proto.GenerateEnvelopeRequest) (*proto.GenerateEnvelopeResponse, error) {
 	var resp proto.GenerateEnvelopeResponse
-	err := run(ctx, p.name, p.path, executor, req, &resp)
+	err := run(ctx, p.name, p.path, req, &resp)
 	return &resp, err
 }
 
 func (p *CLIPlugin) VerifySignature(ctx context.Context, req *proto.VerifySignatureRequest) (*proto.VerifySignatureResponse, error) {
 	var resp proto.VerifySignatureResponse
-	err := run(ctx, p.name, p.path, executor, req, &resp)
+	err := run(ctx, p.name, p.path, req, &resp)
 	return &resp, err
 }
 
-// NewCLIPlugin validate the metadata of the plugin and return a *CLIPlugin.
-func NewCLIPlugin(name, path string) (*CLIPlugin, error) {
-	plugin := CLIPlugin{
-		name: name,
-		path: path,
-	}
-
-	// validate metadata
-	metadata, err := plugin.GetMetadata(context.Background(), &proto.GetMetadataRequest{})
-	if err != nil {
-		return nil, fmt.Errorf("failed to fetch metadata: %w", err)
-	}
-	if metadata.Name != name {
-		return nil, fmt.Errorf("executable name must be %q instead of %q", binName(metadata.Name), filepath.Base(path))
-	}
-	if err = metadata.Validate(); err != nil {
-		return nil, fmt.Errorf("invalid metadata: %w", err)
-	}
-
-	return &plugin, nil
-}
-
-func run(ctx context.Context, pluginName string, pluginPath string, cmder commander, req proto.Request, resp interface{}) error {
+func run(ctx context.Context, pluginName string, pluginPath string, req proto.Request, resp interface{}) error {
 	// serialize request
 	data, err := json.Marshal(req)
 	if err != nil {
-		return pluginErr(pluginName, fmt.Errorf("failed to marshal request object: %w", err))
+		return fmt.Errorf("%s: failed to marshal request object: %w", pluginName, err)
 	}
 
 	// execute request
-	out, err := cmder.Output(ctx, pluginPath, string(req.Command()), data)
+	out, err := executor.Output(ctx, pluginPath, req.Command(), data)
 	if err != nil {
 		var re proto.RequestError
 		err = json.Unmarshal(out, &re)
 		if err != nil {
-			return proto.RequestError{Code: proto.ErrorCodeGeneric, Err: fmt.Errorf("error: %v. stderr: %s", err, out)}
+			return proto.RequestError{Code: proto.ErrorCodeGeneric, Err: err}
```

Comment:
There is bug here. What if `Output()` itself fails? such as permission failure, I/O failure...

Could you create an issue to track it and address it in another PR?

Reply:
Replace cmd.Output() with cmd.Run().
cmd.Output() wraps the stderr into the error, but for our case, it is not convenient because we need to get the stderr directly and parse it to built-in type `proto.RequestError`.

Added test cases:
1. OK response
2. . plugin error in valid JSON format
3. System error
4.  plugin error in invalid JSON format


> 3. and 4. share the same code logic, if the stderr cannot be parsed in JSON format, combine the exit error and stderr together to be a new error.